### ISSUE TITLE
[TanStack Start] HMR is not working from root route (__root)

### ISSUE THREAD

**Issue by iamchanii** [TanStack/router#1992]
### Describe the bug

HMR is not working in root route. only working from normal routes.

### Your Example Website or App

https://github.com/TanStack/router/tree/main/examples/react/start-basic-rsc

### Steps to Reproduce the Bug or Issue

1. Clone (or copy or degit) [start-basic-rsc example](https://github.com/TanStack/router/tree/main/examples/react/start-basic-rsc), install dependencies and run dev server.
2. Open http://localhost:3000
3. Try to edit from `__root.tsx`. I just added a link beside Home.

### Expected behavior

In terminal, dev server printed a log about page has reloaded.
```
[vite] page reload app/routes/__root.tsx
```

but the changes are not reflected in the browser.

### Screenshots or Videos

https://github.com/user-attachments/assets/26aaa1a3-80c1-40af-8dd8-f2403051519e

### Platform

- OS: macOs
- Browser: Arc / Chrome
- Version: 126.0.6478.127

**Comment by anulman** [comment]
I'm experiencing this as well, not just in my root route but in any top-level route.

I also have not been able to separate out `components/Foo.tsx` files and have them hot-reload. The dev console & JS console confirm the HMR event has been emitted and received, but the React JS—even for route sub-components modified in non-route files (e.g. `import { FooPage } from '~/components/FooPage';`) does not re-render the HMR'ed component.

**Comment by ChrisEdgington** [comment]
We just added TanStouch router to our vite-based project, and now we are also experiencing this HMR failure. I tried a few different routes, it does seem to only be happening with __root.tsx.

**Comment by KiwiKilian** [comment]
To summarize, we see the following HMR problems with file based routing:

### HMR is not triggered when editing
- `routes/__root.tsx`
- `components/Example.tsx` imported into any `routes/**.tsx` file
- `routes/_layout.tsx`

### HMR is triggered when editing
- `routes/index.tsx`
- `routes/example.example.tsx`  (@anulman are you sure this doesn't work for any plain route (no `_` prefix) file?)
- `routes/with-directory/example.tsx`


All of this can be testet with the [Start Basic StackBlitz](https://stackblitz.com/github/tanstack/router/tree/main/examples/react/start-basic?embed=1&theme=light&preset=node&file=src/main.tsx).

**Comment by rin-yato** [comment]
Having the same problem for `routes/page-a/-component/example.tsx`

**Comment by jwtong** [comment]
Bump, this happens on routes at random

**Comment by schiller-manuel** [comment]
can you please test out the package built for this PR: https://github.com/TanStack/router/pull/2286

**Comment by jwtong** [comment]
> can you please test out the package built for this PR whether it fixes the HMR issues you are experiencing: [#2286 (comment)](https://github.com/TanStack/router/pull/2286#issuecomment-2336435987)

`    "@tanstack/router": "github:lithdew/router#main"`

Installed as above, still had the same issue though. Updates to pages don't hot reload.

**Comment by jwtong** [comment]
Hmm, yeah looks like I'm not sure how to install your forked version to @tanstack/react-router.

**Comment by KiwiKilian** [comment]
@schiller-manuel I've tried both #2286 and #2316. Both seem to fix the problem in their current state.

@jwtong make sure to unfold the instructions in the comments on how to install [from a pr](https://github.com/TanStack/router/pull/2316#issuecomment-2341913630). You can either use `install`/`add` with your prefered package manager or change the version in your `package.json` to the URL provided. E.g. testing the #2316 looks like this in our `package.json:

```json
    "@tanstack/react-router": "https://pkg.pr.new/@tanstack/react-router@2316",
    "@tanstack/react-router-with-query": "https://pkg.pr.new/@tanstack/react-router-with-query@2316",
    "@tanstack/start": "https://pkg.pr.new/@tanstack/start@2316",
```

Make sure all `@tanstack` packages use the same "version" (rather now PR URL).

**Comment by jwtong** [comment]
@KiwiKilian @schiller-manuel 

Tested both https://github.com/TanStack/router/pull/2286 and https://github.com/TanStack/router/pull/2316.

For me, it did not resolve the issue. If I edit a route file (e.g. `apps/web/src/routes/_auth._nav.member.clients.index.tsx`), I'm still not seeing updates until I refresh the page.

**Comment by KiwiKilian** [comment]
It works for me on #2316 (version `1.58.5` and hash `684eef6`) when editing `app/routes/_authenticated.index.tsx` like just adding some text or changing.

**Comment by pedrobzz** [comment]
none of those fixed my issue. I'm not using tanstack start, only tanstack router, but i have the exact same problem

**Comment by MarkSFrancis** [comment]
For anyone reading this who's looking for a workaround until one of the fixes are merged, you can add the following to your `__root.tsx` to enable HMR in Tanstack Start:
```ts
export const Route = createRootRoute({
  // Other route props like the component
  // ...
  //
  scripts: () =>
    import.meta.env.DEV
      ? [
          {
            type: 'module',
            children: `import RefreshRuntime from "/_build/@react-refresh";
RefreshRuntime.injectIntoGlobalHook(window)
window.$RefreshReg$ = () => {}
window.$RefreshSig$ = () => (type) => type`,
          },
        ]
      : [],
});
```

**Comment by cameronb23** [comment]
@MarkSFrancis Lifesaver, been trying to work Start into a side project of mine and was pretty sad about the HMR (but its alpha so who cares). A great fix for now until the team gets it going!

**Comment by aaronksaunders** [comment]
@MarkSFrancis - worked wonders for me trying to get authentication working, as soon as i added useServerFn everything exploded!! Found this issue and things are rolling again

**Comment by dmagunov** [comment]
Hi, I'm experiencing the same issue with TanStack/Router. HMR doesn't working when I make changes in `__root.tsx`. I've tried various combinations of TanStack/Router (version 1.78.3 and below) with Vite (version 5.4.10 and below) using React/React SWC, but nothing has worked. I also tried @MarkSFrancis's workaround without success.

**Comment by dmagunov** [comment]
Extracting root route component fixed an issue (thanks Claude):

From:
```javascript
createRootRoute({
  component: () => (<></>)
});
```

To:
```javascript
createRootRoute({
  component: RootComponent
});
```

**Comment by dh94** [comment]
> Extracting root route component fixed an issue (thanks Claude):
> 
> From:
> 
> createRootRoute({
>   component: () => (<></>)
> });
> To:
> 
> createRootRoute({
>   component: RootComponent
> });

For me this was the issue in all routes! bless you

**Comment by notKamui** [comment]
Currently, while the page is indeed reloaded upon file modification, it does a full page reload instead of a HMR. But this only happens on regular routes, not the __root route (on which HMR works). Using Tanstack Start

**Comment by vKongv** [comment]
> Currently, while the page is indeed reloaded upon file modification, it does a full page reload instead of a HMR. But this only happens on regular routes, not the __root route (on which HMR works). Using Tanstack Start

I am experiencing full reload too. Not sure what am i missing. Using Start

**Comment by yassine-safraoui** [comment]
> For anyone reading this who's looking for a workaround until a fix is merged, you can add the following to your `__root.tsx` to enable HMR in Tanstack Start:
> 
> export const Route = createRootRoute({
>   // Other route props like the component
>   // ...
>   //
>   scripts: () =>
>     import.meta.env.DEV
>       ? [
>           {
>             type: 'module',
>             children: `import RefreshRuntime from "/_build/@react-refresh";
> RefreshRuntime.injectIntoGlobalHook(window)
> window.$RefreshReg$ = () => {}
> window.$RefreshSig$ = () => (type) => type`,
>           },
>         ]
>       : [],
> });
> This does the same thing as [#2286](https://github.com/TanStack/router/pull/2286) and [#2316](https://github.com/TanStack/router/pull/2316) (albeit in a slightly more hacky way). Once the fix is merged, you should be able to safely remove this code from your `__root.tsx` again

I get type errors when i try to add the script like you did, I tried doing the same thing by replacing scripts with loader or beforeReload but to no avail. i'm using tanstack router( not start) with vite, when I edit and save a component in a file that starts with "-"( inside my routes folder to ignore it) I don't see any changes, not even a full page reload let alone HMR. This is super annoying I guess I can't continue with tanstack router for the time being unfortunately :((

**Comment by shestaya-liniya** [comment]
@yassine-safraoui same :(

**Comment by sngmn451** [comment]
`app/routes/__root.tsx`

```
...
        {import.meta.env.DEV && (
          <script
            type="module"
            dangerouslySetInnerHTML={{
              __html: `import RefreshRuntime from "/_build/@react-refresh";
      RefreshRuntime.injectIntoGlobalHook(window)
      window.$RefreshReg$ = () => {}
      window.$RefreshSig$ = () => (type) => type`
            }}
          />
        )}
</head>
```
This work for me

**Comment by shestaya-liniya** [comment]
@sngmn451 thank you! worked for me too

**Comment by yassine-safraoui** [comment]
> `app/routes/__root.tsx`
> 
> <head>
> ...
>   {import.meta.env.DEV && (
>     <script
>       type="module"
>       dangerouslySetInnerHTML={{
>         __html: `import RefreshRuntime from "/_build/@react-refresh";
> RefreshRuntime.injectIntoGlobalHook(window)
> window.$RefreshReg$ = () => {}
> window.$RefreshSig$ = () => (type) => type`
>       }}
>     />
>   )}
> </head>
> This work for me

not for me :((
doing it like this:
`function RootComponent() {
  return (
    <>
      <head>
        <script
          type="module"
          dangerouslySetInnerHTML={{
            __html: `import RefreshRuntime from "/_build/@react-refresh";
        RefreshRuntime.injectIntoGlobalHook(window)
        window.$RefreshReg$ = () => {}
        window.$RefreshSig$ = () => (type) => type`,
          }}
        />
      </head>
      <Outlet />;
    </>
  );
}`
 gives this error: <head> cannot appear as a child of <div>. it does because at the end all of the stuff in __root goes in the root div in index.html, if I remove the head and keep just the script, it does not work, if I try to add the script manually to the head in the index.html I get errors because I can't import there, even when specifying script type to module.

**Comment by timbo-tj** [comment]
@MarkSFrancis  Thanks!

It didn't work for me at first. I had to marked my Route components for `export` and now it works!

ie instead of:

```typescript
export const Route = createFileRoute('/')({
     component: RouteComponent,
   ...
})

function RouteComponent() {
   ...
}
```

I had to do:

```typescript
export const Route = createFileRoute('/')({
     component: RouteComponent,
   ...
})

export function RouteComponent() {
   ...
}
```

**Comment by notKamui** [comment]
@timbo-tj I don't think this is what you actually want. I works for the wrong reason: Vite cannot do HMR when you export components and non components in the same file. When you do so, it will fallback to full page reloads, which is why you think reload works. I mean, it does, but it's not HMR.

**Comment by timbo-tj** [comment]
@notKamui  Hmmm, I won't claim to understand what Vite + HMR is doing or why...! but this is my experience:

https://github.com/user-attachments/assets/cd1c3df2-71bc-429d-988a-678d44ab38fd

It doesn't seem like a full page reload. (I need to search for my app again on each page load).

**Comment by PhilipGrefe** [comment]
FYI I'm experiencing the same issue (all expect the root routes perform a full reload) when enabling autoCodeSplitting in my tsr.config.json 🫤
```json
{
  "autoCodeSplitting": true
}
```

**Comment by mcintyret** [comment]
Any updates on when https://github.com/TanStack/router/pull/2316 will merge?

Will this also fix HMR when using SWC?

**Comment by SoloReverse** [comment]
My problem was there was no HMR in _components_ only but it was working in routes. I used @MarkSFrancis 's Fix but had to remove the import. this is what worked for me was, I added the following to the head without the ```import.meta.env.dev```. figured I'll just comment it out from builds but it works. 


```
    scripts: [
      {
        type: "module",
        children: `import RefreshRuntime from "/_build/@react-refresh";
    RefreshRuntime.injectIntoGlobalHook(window)
    window.$RefreshReg$ = () => {}
    window.$RefreshSig$ = () => (type) => type`,
      },
    ],
```

FYI I'm using TanStack Start

**Comment by jonathanbecerra** [comment]
I was pulling my hair to why I had to manually refresh the browser manually. My `package.json` has the following:
```json
```

**Comment by jonathanbecerra** [comment]
I was pulling my hair and wondering why my changes outside of the `routes/` but still under `app/` are not properly watched. 
I find it odd that vite's hrm alludes that it did update when it didn't. Main reason for the hair pulling. Example of my console:
```bash
4:54:13 PM [vite] (client) hmr update /app/components/SideBarMenu.tsx, /app/styles/app.css?direct
```
My working directory as followed:
```bash
app
├── api.tsx
├── client.tsx
├── components
│   ├── SideBarMenu.tsx // -> This does not update ❌
├── routes // -> Anything under routes does update ✅
│   ├── __root.tsx 
```
My `package.json' has the following:
```json
    "@tanstack/react-query": "^5.62.10",
    "@tanstack/react-query-devtools": "^5.62.10",
    "@tanstack/react-router": "^1.92.3",
    "@tanstack/react-router-with-query": "^1.92.3",
    "@tanstack/router-devtools": "^1.92.3",
    "@tanstack/router-plugin": "^1.91.1",
    "@tanstack/start": "^1.92.4",
```
The workaround on this thread that adds a script tag injecting the refresh runtime will throw an error if added to the `__root.tsx` or anywhere else. The error thrown in the browser is the following:
```ts
[plugin:vite-plugin-tanstack-start-create-server-fn] We detected that the '@vitejs/plugin-react' was passed before '@tanstack/start-vite-plugin'. Please make sure that '@tanstack/router-vite-plugin' is passed before '@vitejs/plugin-react' and try again: 
e.g.

plugins: [
  TanStackStartVite(), // Place this before viteReact()
  viteReact(),
]
app/routes/__root.tsx
    at TransformPluginContext.transform (file://node_modules/.pnpm/@tanstack+start-vite-plugin@1.91.1/node_modules/@tanstack/start-vite-plugin/dist/esm/index.js:33:15)
    at EnvironmentPluginContainer.transform (file://node_modules/.pnpm/vite@6.0.6_@types+node@22.10.2_jiti@2.4.2_terser@5.37.0_tsx@4.19.2_yaml@2.6.1/node_modules/vite/dist/node/chunks/dep-0AosnpPU.js:47220:19)
    at async loadAndTransform (file://node_modules/.pnpm/vite@6.0.6_@types+node@22.10.2_jiti@2.4.2_terser@5.37.0_tsx@4.19.2_yaml@2.6.1/node_modules/vite/dist/node/chunks/dep-0AosnpPU.js:41030:27)
    at async viteTransformMiddleware (file://node_modules/.pnpm/vite@6.0.6_@types+node@22.10.2_jiti@2.4.2_terser@5.37.0_tsx@4.19.2_yaml@2.6.1/node_modules/vite/dist/node/chunks/dep-0AosnpPU.js:42474:24
Click outside, press Esc key, or fix the code to dismiss.
You can also disable this overlay by setting server.hmr.overlay to false in vite.config.js.
```
Any new suggestions will be happy to add to test. For now we keep manually `⌘ + r`. Happy holidays

**Comment by menisck** [comment]
> I was pulling my hair and wondering why my changes outside of the `routes/` but still under `app/` are not properly watched. I find it odd that vite's hrm alludes that it did update when it didn't. Main reason for the hair pulling. Example of my console:
> 
> 4:54:13 PM [vite] (client) hmr update /app/components/SideBarMenu.tsx, /app/styles/app.css?direct
> My working directory as followed:
> 
> app
> ├── api.tsx
> ├── client.tsx
> ├── components
> │   ├── SideBarMenu.tsx // -> This does not update ❌
> ├── routes // -> Anything under routes does update ✅
> │   ├── __root.tsx 
> My `package.json' has the following:
> 
>     "@tanstack/react-query": "^5.62.10",
>     "@tanstack/react-query-devtools": "^5.62.10",
>     "@tanstack/react-router": "^1.92.3",
>     "@tanstack/react-router-with-query": "^1.92.3",
>     "@tanstack/router-devtools": "^1.92.3",
>     "@tanstack/router-plugin": "^1.91.1",
>     "@tanstack/start": "^1.92.4",
> The workaround on this thread that adds a script tag injecting the refresh runtime will throw an error if added to the `__root.tsx` or anywhere else. The error thrown in the browser is the following:
> 
> [plugin:vite-plugin-tanstack-start-create-server-fn] We detected that the '@vitejs/plugin-react' was passed before '@tanstack/start-vite-plugin'. Please make sure that '@tanstack/router-vite-plugin' is passed before '@vitejs/plugin-react' and try again: 
> e.g.
> 
> plugins: [
>   TanStackStartVite(), // Place this before viteReact()
>   viteReact(),
> ]
> app/routes/__root.tsx
>     at TransformPluginContext.transform (file://node_modules/.pnpm/@tanstack+start-vite-plugin@1.91.1/node_modules/@tanstack/start-vite-plugin/dist/esm/index.js:33:15)
>     at EnvironmentPluginContainer.transform (file://node_modules/.pnpm/vite@6.0.6_@types+node@22.10.2_jiti@2.4.2_terser@5.37.0_tsx@4.19.2_yaml@2.6.1/node_modules/vite/dist/node/chunks/dep-0AosnpPU.js:47220:19)
>     at async loadAndTransform (file://node_modules/.pnpm/vite@6.0.6_@types+node@22.10.2_jiti@2.4.2_terser@5.37.0_tsx@4.19.2_yaml@2.6.1/node_modules/vite/dist/node/chunks/dep-0AosnpPU.js:41030:27)
>     at async viteTransformMiddleware (file://node_modules/.pnpm/vite@6.0.6_@types+node@22.10.2_jiti@2.4.2_terser@5.37.0_tsx@4.19.2_yaml@2.6.1/node_modules/vite/dist/node/chunks/dep-0AosnpPU.js:42474:24
> Click outside, press Esc key, or fix the code to dismiss.
> You can also disable this overlay by setting server.hmr.overlay to false in vite.config.js.
> Any new suggestions will be happy to add to test. For now we keep manually `⌘ + r`. Happy holidays

I'm facing the same issue...

**Comment by lFitzl** [comment]
I had to do this to get it to work....

![Image](https://github.com/user-attachments/assets/ee74d2e0-036f-4725-91c2-618863be7367)

**Comment by forfalsket** [comment]
> [@MarkSFrancis](https://github.com/MarkSFrancis) Thanks!
> 
> It didn't work for me at first. I had to marked my Route components for `export` and now it works!
> 
> ie instead of:
> 
> export const Route = createFileRoute('/')({
>      component: RouteComponent,
>    ...
> })
> 
> function RouteComponent() {
>    ...
> }
> I had to do:
> 
> export const Route = createFileRoute('/')({
>      component: RouteComponent,
>    ...
> })
> 
> export function RouteComponent() {
>    ...
> }

as soon as I used functions instead of const, it worked. You're god's send!

**Comment by masnormen** [comment]
Thanks, @MarkSFrancis 's solution works. But the value of `script` should be a regular array, not a function that returns array. Otherwise I get type errors.

Here's a version that works with `"@tanstack/react-router": "^1.95.3"` and `"@tanstack/start": "^1.95.3"`:

```ts
// __root.tsx

export const Route = createRootRoute({
  // .....
  scripts:
    import.meta.env.DEV
      ? [
          {
            type: 'module',
            children: `import RefreshRuntime from "/_build/@react-refresh";
RefreshRuntime.injectIntoGlobalHook(window)
window.$RefreshReg$ = () => {}
window.$RefreshSig$ = () => (type) => type`,
          },
        ]
      : [],
});
```

**Comment by Aias** [comment]
@masnormen This is the only solution that's worked for me so far, amazing finally having hot reloading work for everything without having to manually refresh the page on every change.

Not sure if this is the same for everyone else experiencing the issue, but for me hot reloading wasn't working basically anywhere outside of main Route components, even for subcomponents within the same route file. Changes in any directory besides `app/routes` wouldn't get picked up at all until a hard page refresh. So this is a great workaround until that can get fixed, surprised it's not higher priority considering the DX benefits.

**Comment by yassine-safraoui** [comment]
> Thanks, [@MarkSFrancis](https://github.com/MarkSFrancis) 's solution works. But the value of `script` should be a regular array, not a function that returns array. Otherwise I get type errors.
> 
> Here's a version that works with `"@tanstack/react-router": "^1.95.3"` and `"@tanstack/start": "^1.95.3"`:
> 
> // __root.tsx
> 
> export const Route = createRootRoute({
>   // .....
>   scripts:
>     import.meta.env.DEV
>       ? [
>           {
>             type: 'module',
>             children: `import RefreshRuntime from "/_build/@react-refresh";
> RefreshRuntime.injectIntoGlobalHook(window)
> window.$RefreshReg$ = () => {}
> window.$RefreshSig$ = () => (type) => type`,
>           },
>         ]
>       : [],
> });

how are you guys able to provide a scripts array to the options of createRootRoute, typescript tells me 'scripts' does not exist in type RootRouteOptions, I'm using tanstack router and tried updating it to 1.95.3 but to no avail.
```typescript
import { Outlet, createRootRoute } from "@tanstack/react-router";

export const Route = createRootRoute({
  component: RootComponent,
  // I get a ts error here saying: Object literal may only specify known properties, and 'scripts' does not exist in type 'RootRouteOptions<undefined, {}, AnyContext, AnyContext, {}, undefined>'.ts(2353)
  scripts: import.meta.env.DEV
    ? [
        {
          type: "module",
          children: `import RefreshRuntime from "/_build/@react-refresh";
RefreshRuntime.injectIntoGlobalHook(window)
window.$RefreshReg$ = () => {}
window.$RefreshSig$ = () => (type) => type`,
        },
      ]
    : [],
});

export function RootComponent() {
  return (
    <>
      <Outlet />;
    </>
  );
}
```

**Comment by masnormen** [comment]
@yassine-safraoui I mixed it up. It should go inside the return value of the `head` property's function. Just updated my comment

**Comment by yassine-safraoui** [comment]
> [@yassine-safraoui](https://github.com/yassine-safraoui) I messed it up. It should go inside the return value of the `head` property's function. Just updated my comment

okay thanks, I just tried it now and it's still not working 🫤. I litereally copy pasted the head you have into my root, I made sure i'm exporting the RootComponent, I even tried moving it into its own file(-RootComponent.tsx), the page never hot reloaded when I changed a component inside RouteComponent.

**Comment by jonathanbecerra** [comment]
> import.meta.env.DEV
>       ? [
>           {
>             type: 'module',
>             children: `import RefreshRuntime from "/_build/@react-refresh";
> RefreshRuntime.injectIntoGlobalHook(window)
> window.$RefreshReg$ = () => {}
> window.$RefreshSig$ = () => (type) => type`,
>           },
>         ]
>       : [],

@masnormen thanks, this actually solved my issue. I know the community seems to have some reoccurring instances but for my case, good. To give clarity I have kept up to date with recent releases to make sure Im current. Posting versions in my project:
```json
{
  "@tanstack/react-query": "^5.64.1",
  "@tanstack/react-query-devtools": "^5.64.1",
  "@tanstack/react-router": "^1.97.3",
  "@tanstack/react-router-with-query": "^1.97.3",
  "@tanstack/router-devtools": "^1.97.3",
  "@tanstack/router-plugin": "^1.97.3",
  "@tanstack/start": "^1.97.3"
}
```

**Comment by SeanCassiere** [comment]
The HMR fixes by @lithdew have been merged in and most of the HMR issues should be resolved with the https://github.com/TanStack/router/releases/tag/v1.99.3.

If more, HMR bugs pop up, please open a new Issue on github so that it can be tracked separately.

**Comment by mahmingtea** [comment]
this is strange, whenever i use hooks inside __root component HMR breaks, but when i move these hooks in different component, and import this same component inside __root component, HMR works. __root component does not like it when i use return inside it

**Comment by raves21** [comment]
im still experiencing this issue on a fresh tanstack router project.

I did @masnormen 's fix but it gives me an error:

```
[plugin:tanstack-router:code-splitter:compile-reference-file] We detected that the '@vitejs/plugin-react' was passed before '@tanstack/router-plugin/vite'. Please make sure that '@tanstack/router-plugin' is passed before '@vitejs/plugin-react' and try again: 
e.g.
plugins: [
  tanstackRouter(), // Place this before viteReact()
  viteReact(),
]
```

But this is my vite.config.ts:

```
import { defineConfig } from "vite";
import { tanstackRouter } from "@tanstack/router-plugin/vite";
import react from "@vitejs/plugin-react-swc";
import tailwindcss from "@tailwindcss/vite";
import path from "path";

// https://vite.dev/config/
export default defineConfig({
  plugins: [
    tanstackRouter({
      target: "react",
      autoCodeSplitting: true,
    }),
    react(),
    tailwindcss(),
  ],
  resolve: {
    alias: {
      "@": path.resolve(__dirname, "./src"),
    },
  },
});
```

**Comment by schiller-manuel** [comment]
@raves21 can you please create a new GitHub issue including a complete minimal example project that reproduces this?

**Comment by raves21** [comment]
> [@raves21](https://github.com/raves21) can you please create a new GitHub issue including a complete minimal example project that reproduces this?

Ok nvm its fixed now I did @timbo-tj 's solution which was make the RootComponent into a named function.

From this:
```
import { createRootRoute, Outlet } from "@tanstack/react-router";

export const Route = createRootRoute({
  component: () => (
    <>
      <div className="flex items-center gap-4">
        <p>Home</p>
        <p>Test</p>
      </div>
      <Outlet />
    </>
  ),
});
```

To this:
```
import { createRootRoute, Outlet } from "@tanstack/react-router";

export const Route = createRootRoute({
  component: RootComponent,
});

function RootComponent() {
  return (
    <>
      <div className="flex items-center gap-4">
        <p>Home</p>
        <p>Test</p>
      </div>
      <Outlet />
    </>
  );
```
}

**Comment by daveycodez** [comment]
This plugin will fix HMR for beforeLoad & editing json files or anything outside of routes directory:

```tsx
import type { Plugin } from "vite"

export function tanstackRouterHMR(): Plugin {
    return {
        name: "tanstack-router-hmr",
        enforce: "post",
        handleHotUpdate(ctx) {
            const invalidatedModules = []

            for (const mod of ctx.server.moduleGraph.idToModuleMap.values()) {
                if (mod.id?.includes("/router.ts")) {
                    invalidatedModules.push(mod)
                }
            }

            return invalidatedModules
        }
    }
}
```

Just add this plugin to your Vite config. Hopefully this can be fixed officially for 1.0.
